### PR TITLE
Startup notification callbacks

### PR DESCRIPTION
**Pull request by Elv13**
This pull finally complete a feature that started in my June 2013 pull. It is now possible to have pass arguments to awful.spawn() to, for example, spawn in a new tag. It also add some other things to make it easier to use.

It is a "big" pull, sorry about that. I tried to avoid making so much changes, but in the end they are required, so umm, have an happy time reviewing this.

### CODE REVIEW

**Comment by Elv13** [comment]
Closing this, its code rot. The necessary commits have long been merged and those who showed interest for the other ones already switched to the Tyrannical implementation as it fit that crowd needs better anyway.

**Comment by blueyed** [comment]
@Elv13 
While it's old, I disagree with that it should be closed.

I am using some own implementation of sn_properties in my rc.lua, but would really like to have support / a reference implementation for this in awesome itself.

The conflicts against master aren't too bad.

File: lib/awful/init.lua [hunk 1]
```diff
@@ -37,6 +37,7 @@ return
     tooltip = require("awful.tooltip");
     ewmh = require("awful.ewmh");
     titlebar = require("awful.titlebar");
+    spawn = require("awful.spawn");
 }
 
 -- vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
```
Comment (blueyed) [inline lib/awful/init.lua]: The file properties were changed.. should not be executable.

File: lib/awful/rules.lua [hunk 2]
```diff
@@ -233,6 +233,13 @@ function rules.execute(c, props, callbacks)
     end
 end
 
+function rules.completed_with_payload_callback(c,props)
+    rules.execute(c,props,type(props.callback) == "function" and
```
Comment (blueyed) [inline lib/awful/rules.lua]: Please add spaces.. :)

File: lib/awful/spawn.lua [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+--- Spawn a program.
+-- See awful.rules.execute for more details
+-- @param cmd The command.
+-- @param sn_rules a property table, false to disable startup-notification.
+-- @param callback A callback function (if the client support startup notifications)
+-- @return The forked PID or an error message
+-- @return The startup notification UID, if the spawn was successful
+function spawn.spawn(cmd, sn_rules, callback)
+    if cmd and cmd ~= "" then
+        local enable_sn = (sn_rules ~= false or callback) and true or true
```
Comment (blueyed) [inline lib/awful/spawn.lua]: s/ and true or true//

**Comment by blueyed** [comment]
:+1:

**Comment by psychon** [comment]
Reopening in the hope that travis picks this up.

The change is fine with me. However, would it make sense to add tests for this "rules thing"?

**Comment by blueyed** [comment]
> Why doesn't travis pick up this PR?

Maybe because it pre-dates the integration of Travis, which then only considers new PRs?! (just a guess)

**Pull request closed without merging**